src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -390,6 +390,57 @@ private static bool CheckConstraintsSingleType(TypeSymbol type, CheckConstraints
         }
 
         public static bool CheckConstraints(
+            this TupleTypeSymbol tuple,
+            ConversionsBase conversions,
+            SyntaxNode typeSyntax,
+            ImmutableArray<Location> elementLocations,
+            Compilation currentCompilation,
+            DiagnosticBag diagnostics)
+        {
+            NamedTypeSymbol type = tuple.TupleUnderlyingType;
+            if (!RequiresChecking(type))
+            {
+                return true;
+            }
+
+            var diagnosticsBuilder = ArrayBuilder<TypeParameterDiagnosticInfo>.GetInstance();
+            var useSiteDiagnosticsBuilder = (ArrayBuilder<TypeParameterDiagnosticInfo>)null;
+
+            var result = !typeSyntax.HasErrors;
+
+            var underlyingTupleTypeChain = ArrayBuilder<NamedTypeSymbol>.GetInstance();
+            TupleTypeSymbol.GetUnderlyingTypeChain(type, underlyingTupleTypeChain);
+
+            var offset = 0;
+            foreach (var underlyingTuple in underlyingTupleTypeChain)
+            {
+                result &= CheckTypeConstraints(underlyingTuple, conversions, currentCompilation, diagnosticsBuilder, ref useSiteDiagnosticsBuilder);
+
+                if (useSiteDiagnosticsBuilder != null)
+                {
+                    diagnosticsBuilder.AddRange(useSiteDiagnosticsBuilder);
+                }
+
+                foreach (var pair in diagnosticsBuilder)
+                {
+                    var ordinal = pair.TypeParameter.Ordinal + offset;
+                    var location = ordinal < elementLocations.Length ? elementLocations[ordinal] : typeSyntax.Location;
+                    diagnostics.Add(new CSDiagnostic(pair.DiagnosticInfo, location));
+                }
+
+                useSiteDiagnosticsBuilder = null;
+                diagnosticsBuilder.Clear();
+
+                offset += TupleTypeSymbol.RestIndex;
+            }
+
+            underlyingTupleTypeChain.Free();
+            diagnosticsBuilder.Free();
+
+            return result;
+        }
+
+        public static bool CheckConstraints(
```

Comment:
It's always a tuple type as it accepts a `TupleTypeSymbol`. so no assertion is required.